### PR TITLE
Refactor start script and support start script v2

### LINKED ISSUE #4393: runmode=debug support for TiFlash pods
## Feature Request

**Is your feature request related to a problem? Please describe:**

**Describe the feature you'd like:**

`runmode=debug` support for TiFlash pods. https://docs.pingcap.com/zh/tidb-in-kubernetes/stable/tips/#%E8%AF%8A%E6%96%AD%E6%A8%A1%E5%BC%8F

**Describe alternatives you've considered:**

**Teachability, Documentation, Adoption, Migration Strategy:**

### PR DESCRIPTION
**Pull request by KanShiori**
### What problem does this PR solve?

Close #4393
### What is changed and how does it work?

* create pkg `pkg/manager/member/startscript/v2`, and mv `template.go` to pkg `pkg/manager/member/startscript/v1`
* separate `render` func to `pkg/manager/member/startscript`
* add `startScriptVersion` in tc spec to choose startscript

### Code changes

- [x] Has Go code change
- [ ] Has CI related scripts change

### Tests

- [x] Unit test 
- [x] E2E test 
- [ ] Manual test 

Deploy new cluster:
* deploy cluster with `startScriptVersion:v1`
* deploy cluster with `startScriptVersion:v1` and tls-enabled
* deploy cluster with `startScriptVersion:v2` 
* deploy cluster with `startScriptVersion:v2 and tls-enabled

For compatibility:
* deploy cluster by v1.3.2 operator, then upgrade it, check that cluster have not rolling upgrade.
* deploy cluster by v1.3.2 operator, then upgrade it and use `startScriptVersion:v1`,  check that the cluster have not rolling upgrade.
* deploy cluster by v1.3.2 operator, then use `startScriptVersion:v2`,  wait for cluster to be ready and check that the args of all components have not changed.

Debug mod:
* Deploy cluster with `startScriptVersion:v2`, and try to enable debug mod for all components.

- [ ] No code 

### Side effects

- [ ] Breaking backward compatibility 
- [ ] Other side effects: 

### Related changes

- [ ] Need to cherry-pick to the release branch 
- [ ] Need to update the documentation 

### Release Notes

Please refer to [Release Notes Language Style Guide](https://github.com/pingcap/tidb-operator/blob/master/docs/release-note-guide.md) before writing the release note.

```release-note
Add `startScriptVersion` field to choose startscript for all components 
```

### CODE REVIEW

File: pkg/manager/member/tikv_member_manager.go [hunk 1]
```diff
@@ -705,27 +702,7 @@ func getTikVConfigMap(tc *v1alpha1.TidbCluster) (*corev1.ConfigMap, error) {
 		return nil, nil
 	}
 
-	scriptModel := &TiKVStartScriptModel{
```
Comment (KanShiori) [inline pkg/manager/member/tikv_member_manager.go:708]: move to the func `v1.RenderTiKVStartScript`

File: pkg/manager/member/pd_member_manager.go [hunk 2]
```diff
@@ -856,19 +854,7 @@ func getPDConfigMap(tc *v1alpha1.TidbCluster) (*corev1.ConfigMap, error) {
 		return nil, err
 	}
 
-	sm := &PDStartScriptModel{
```
Comment (KanShiori) [inline pkg/manager/member/pd_member_manager.go:859]: move to the func `v1.RenderPDStartScript`

File: pkg/manager/member/pump_member_manager.go [hunk 3]
```diff
@@ -511,54 +515,6 @@ func getPumpMeta(tc *v1alpha1.TidbCluster, nameFunc func(string) string) (metav1
 	return objMeta, pumpLabel
 }
 
-func getPumpStartScript(tc *v1alpha1.TidbCluster) (string, error) {
```
Comment (KanShiori) [inline pkg/manager/member/pump_member_manager.go:514]: move to the function `v1.RenderPumpStartScript`

File: pkg/manager/member/tiflash_member_manager.go [hunk 4]
```diff
@@ -412,30 +413,11 @@ func getNewStatefulSet(tc *v1alpha1.TidbCluster, cm *corev1.ConfigMap) (*apps.St
 			},
 		},
 	}
-	script := "set -ex;ordinal=`echo ${POD_NAME} | awk -F- '{print $NF}'`;sed s/POD_NUM/${ordinal}/g /etc/tiflash/config_templ.toml > /data0/config.toml;sed s/POD_NUM/${ordinal}/g /etc/tiflash/proxy_templ.toml > /data0/proxy.toml"
```
Comment (KanShiori) [inline pkg/manager/member/tiflash_member_manager.go:415]: move into the func `v1.RenderTiFlashInitScript`

**Comment by DanielZhangQD** [comment]
@KanShiori 
Did you check that debug mode works as expected for each supported component with v1 and v2 after this PR?

**Comment by KanShiori** [comment]
> @KanShiori Did you check that debug mode works as expected for each supported component with v1 and v2 after this PR?

Yes, I did

File: pkg/manager/member/startscript/v2/pd_start_script.go [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+}
+
+// RenderPDStartScript renders PD start script from TidbCluster
+func RenderPDStartScript(tc *v1alpha1.TidbCluster) (string, error) {
+	m := &PDStartScriptModel{}
+	tcName := tc.Name
+	tcNS := tc.Namespace
+	peerServiceName := controller.PDPeerMemberName(tcName)
+
+	m.PDDomain = fmt.Sprintf("${PD_POD_NAME}.%s.%s.svc", peerServiceName, tcNS)
```
Comment (DanielZhangQD) [inline pkg/manager/member/startscript/v2/pd_start_script.go:46]: Is it ok to remove `.svc` if no clusterDomain is specified to reduce the useless dns query?
And if we do this, the code in the discovery [here](https://github.com/pingcap/tidb-operator/blob/v1.3.2/pkg/discovery/discovery.go#L82) also needs to update.

Comment (DanielZhangQD) [reply pkg/manager/member/startscript/v2/pd_start_script.go:46]: Please also check for the other components.

Comment (DanielZhangQD) [reply pkg/manager/member/startscript/v2/pd_start_script.go:46]: OK, I see, this is to keep the backward compatibility with v1 start scripts.

File: tests/e2e/tidbcluster/tidbcluster.go [hunk 6]
```diff
@@ -2999,6 +2999,132 @@ var _ = ginkgo.Describe("TiDBCluster", func() {
...
+				tlsEnable:  false,
+			},
+			{
+				nameSuffix: "and enable TLS",
+				tlsEnable:  true,
+			},
+		}
+
+		for _, testcase := range cases {
+			ginkgo.It("deploy cluster with start script v2 "+testcase.nameSuffix, func() {
```
Comment (DanielZhangQD) [inline tests/e2e/tidbcluster/tidbcluster.go:3021]: We need to add similar two cases for across Kubernetes deployment.

Comment (KanShiori) [reply tests/e2e/tidbcluster/tidbcluster.go:3021]: Done https://github.com/pingcap/tidb-operator/pull/4505/commits/42f2b0f1ba0aa938589b94a72cfdbe547545f012

File: pkg/apis/pingcap/v1alpha1/tidbcluster.go [hunk 7]
```diff
@@ -964,3 +979,12 @@ func (tc *TidbCluster) WithoutLocalTiDB() bool {
 func (tc *TidbCluster) AcrossK8s() bool {
 	return tc.Spec.AcrossK8s
 }
+
+func (tc *TidbCluster) StartScriptVersion() StartScriptVersion {
```
Comment (july2993) [inline pkg/apis/pingcap/v1alpha1/tidbcluster.go:983]: how about:

```go
  func (tc *TidbCluster) StartScriptVersion() StartScriptVersion {
      if tc.Spec.StartScriptVersion == "" {
          return StartScriptV1
      }

      return tc.Spec.StartScriptVersion
  }
```

**Review by july2993 — APPROVED** [review]

**Review by DanielZhangQD — APPROVED** [review]
LGTM

File: pkg/apis/pingcap/v1alpha1/tidbcluster.go (hunk 7 above)
Comment (DanielZhangQD) [reply pkg/apis/pingcap/v1alpha1/tidbcluster.go:983]: For non-supported configurations, the current code will return v1.

File: pkg/manager/member/startscript/v2/ticdc_start_script.go [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+	AcrossK8s *AcrossK8sScriptModel
+}
+
+func RenderTiCDCStartScript(tc *v1alpha1.TidbCluster) (string, error) {
+	m := &TiCDCStartScriptModel{}
+	tcName := tc.Name
+	tcNS := tc.Namespace
+	peerServiceName := controller.TiCDCPeerMemberName(tcName)
+
+	advertiseAddr := fmt.Sprintf("${TICDC_POD_NAME}.%s.%s.svc", peerServiceName, tcNS)
```
Comment (overvenus) [inline pkg/manager/member/startscript/v2/ticdc_start_script.go:47]:
```
	// NB: TiCDC control relies the format.
	// TODO move advertise addr format to package controller.
```

Please add the comments, thanks!

Comment (KanShiori) [reply pkg/manager/member/startscript/v2/ticdc_start_script.go:49]: Done f52f13f

**Pull request merged**